source/speechXml.py

```diff
@@ -280,3 +304,125 @@ def convertVolumeCommand(self, command):
+			prevCommand = next((c for c in reversed(self._speechSequence) if type(c) is type(command)), None)
+			if prevCommand != command:
+				self._speechSequence.append(command)
+
+	def convertFromXml(self, xml: str) -> SpeechSequence:
+		"""Convert XML to a speech sequence.
+		"""
+		self._speechSequence = SpeechSequence()
+		parser = expat.ParserCreate('utf-8')
+		parser.StartElementHandler = parser.EndElementHandler = self._elementHandler
+		parser.CharacterDataHandler = self._speechSequence.append
+		try:
+			parser.Parse(xml)
+		except Exception as e:
+			raise ValueError(f"XML: {xml}") from e
+		return self._speechSequence
+
+
+ParseGeneratorT = Generator[SpeechCommand, None, None]
+ParseFuncT = Callable[[dict[str, str] | None], ParseGeneratorT]
+MarkCallbackT = Callable[[str], None]
+
+
+class SsmlParser(SpeechXmlParser):
+	"""Parses SSML into an NVDA speech sequence.
+	"""
+
+	def __init__(self, markCallback: MarkCallbackT | None = None):
+		"""Constructor.
+
+		:param markCallback: An optional callback called for every mark command in the SSSML.
+			The mark command in the SSML will be translated to a CallbackCommand instance.
+		"""
+		self._markCallback = markCallback
+
+	def parseSayAs(self, attrs: dict[str, str] | None) -> ParseGeneratorT:
+		state = attrs is not None and attrs.get("interpret-as") == "characters"
+		yield CharacterModeCommand(state)
+
+	def parseVoice(self, attrs: dict[str, str] | None) -> ParseGeneratorT:
+		if attrs is None:
+			return
+		if (xmlLang := attrs.get("xml:lang")) is None:
+			return
+		yield LangChangeCommand(toNvdaLang(xmlLang))
+
+	def parseBreak(self, attrs: dict[str, str] | None) -> ParseGeneratorT:
+		if attrs is None or "time" not in attrs:
+			return
+		if (time := RE_TIME_MS.match(attrs["time"])) is None:
+			log.debugWarning(f"Unknown attributes for break tag: {attrs}")
+			return
+		yield BreakCommand(int(time.group("time")))
+
+	_cachedProsodyAttrs: list[dict]
+
+	def parseProsody(self, attrs: dict[str, str] | None) -> ParseGeneratorT:
+		if isOpenTag := attrs is not None:
+			self._cachedProsodyAttrs.append(attrs)
+		else:  # attrs is None
+			# Pop the attrs from the cache so we can add commands to reset them.
+			attrs = self._cachedProsodyAttrs.pop()
+		for attr, val in attrs.items():
+			if (percentage := RE_PERCENTAGE.match(val)) is None:
+				log.debugWarning(f"Attribute {attr!r} for prosody tag has unparseable value: {val!r}")
+				continue
+			multiplier = float(percentage.group("percentage")) / 100 if isOpenTag else 1
+			match attr:
+				case "pitch":
+					yield PitchCommand(multiplier=multiplier)
+				case "volume":
+					yield VolumeCommand(multiplier=multiplier)
+				case "rate":
+					yield RateCommand(multiplier=multiplier)
+				case _:
+					log.debugWarning(f"Unknown prosody attribute: {attr!r}")
+					continue
+
+	def parseSpeak(self, attrs: dict[str, str] | None) -> ParseGeneratorT:
+		return
+		yield

```

Comment:
I want it to yield nothing but it should still be a generator. Returning before yielding ensures that the generator is empty. If we'd only yield, it would be a generator yielding `None` once. When using only return, it wouldn't be a generator.
I'm happy for any suggestions to improve this though, but I figured this was the best way to accomplish this.

Reply:
Ah I wasn't aware of this syntax, thanks